### ISSUE TITLE
Statically Link libc for alpine distribution of `pachctl`

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#2193]
Reported by a user:

```
Trying to download linux distro of pachyderm pachctl_1.5.3_linux_amd64.tar.gz
pachctl won't run on Alpine Docker container due to error code: 127
ldd /usr/local/bin/pachctl
    /lib64/ld-linux-x86-64.so.2 (0x55b1119f7000)
    libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x55b1119f7000)
    libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55b1119f7000)
```

What's unclear to me is how running the worker binary on the workers (w an alpine image) works at all today. I know we've had users run jobs w images based on alpine that work fine. I would expect to see a similar error to the above. I suppose its possible pachctl makes syscalls / has shared lib dependencies the worker binary doesn't, though that seems unlikely. It's also possible we don't have across the board support for all alpine versions.

We should [be able to statically link in libc](https://stackoverflow.com/questions/13187499/link-glibc-statically-but-some-other-library-dynamically-with-gcc). That should 'fix' the error above about resolving libraries at runtime. However, I believe that 'locks us in' to whatever kernels support that version of glibc. (I've seen it before where statically linked binaries are deployed, but the OS dependency needs to be managed to use the same version). I'm not sure which kernel versions Alpine supports.

Additionally, it'll probably make the `pachctl` binary quite a bit larger. In this case we'd want to release a new target that was statically linked.

TL;DR ... it may 'fix' the library resolution problem, but not give us compatibility w all docker image versions (perhaps not even all of alpine).

So ... it may be much more valuable to provide a make task to do the compilation themselves (so any user could compile / statically link against whatever version of glibc they have).

[And in fact, I think the glibc license LGPL means](https://lwn.net/Articles/117972/) we can't statically link and distribute the binary.

I think we could create a make task that would allow you to do this yourself. Or ... support a docker image w the glibc library in there ... that you could link against ... but I suspect this defeats the intent of a lot of people's usage of alpine (and lightweight image sizes).

**Comment by bpb** [comment]
What is the suggestion for the time being? Use a ubuntu container or build my own Pachctl for alpine?

**Comment by sjezewski** [comment]
Yes, an ubuntu image should work for now.

The probably 'todo' from this issue will be to create a make task so that its easy to roll your own. You're welcome to to that of course, but we'd want to make it easier by putting that into a task for you.

I am curious -- it seems like you're trying to use `pachctl` within an alpine container. Can I ask why?

**Comment by bpb** [comment]
Bootstrapping pachyderm into a K8s AWS cluster using a docker container.

**Comment by bpb** [comment]
simply grabbing a `golang:alpine` container and running
`$ CGO_ENABLED=0;go build -ldflags '-extldflags "-static"' ./src/server/cmd/pachctl`
does the trick

**Comment by sjezewski** [comment]
stale
